### PR TITLE
CLN+TST: Catch specific exception in equals

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

**Review by WillAyd — APPROVED** [review]
lgtm. Do you know generally how many more `except Exceptions` are left to tackle?

**Comment by jbrockmendel** [comment]
> Do you know generally how many more except Exceptions are left to tackle?

106 on master.  A handful of those are correct as is, not sure how many (maybe we could alias `AnyException = Exception` and so that grepping for "except Exception" doesn't turn those up?  just brainstorming)

Also a handful of `BaseException` in `test_nanops` that we _really_ shouldn't be catching.  Looks like we got rid of all the `except:`.

**Comment by TomAugspurger** [comment]
Is this a case where we want to be broad in what we catch? Having equality checks never raise is convenient.

**Comment by jbrockmendel** [comment]
> Is this a case where we want to be broad in what we catch?

That's not unreasonable.  I lean towards being specific because it surfaces other bugs (like passing MultiIndex raising the wrong exception).

**Comment by WillAyd** [comment]
I generally dislike `except Exception` - maybe we find some other case we haven't discovered but would also prefer to be explicit with that when it surfaces

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -1918,6 +1918,9 @@ def sequence_to_dt64ns(
     tz = validate_tz_from_dtype(dtype, tz)
 
     if isinstance(data, ABCIndexClass):
+        if data.nlevels > 1:
+            # Without this check, data._data below is None
+            raise TypeError("Cannot create a DatetimeArray from a MultiIndex.")
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py]: we should have an index method which does this no?

Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: For the check or the raising?  The check could be `isinstance(data, ABCMultiIndex)`.  No strong preference.

Comment (jreback) [reply pandas/core/arrays/datetimes.py]: no, something that does the right thing on an Index/MultiIndex, e.g. that gets (data._data) on an index and just passes thru on a MI

Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: I think we want to move to use the `extract_array` pattern here, but that currently raises ValueError on MultiIndex (where we want to raise TypeError here)

Comment (jreback) [reply pandas/core/arrays/datetimes.py]: ok the reason I am harping on this is its an inconsistency between Index and MultiIndex. ok on merging this and creating an issue / followup ?

Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: > creating an issue / followup ?

yes

**Comment by jreback** [comment]
k cool

**Pull request merged**